atomicsabi64/atomicsabi64.rst

```diff
@@ -0,0 +1,1161 @@
+..
+   Copyright (c) 2024, Arm Limited and its affiliates.  All rights reserved.
+   CC-BY-SA-4.0 AND Apache-Patent-License
+   See LICENSE file for details
+
+.. |release| replace:: 2024Q1
+.. |date-of-issue| replace:: 5\ :sup:`th` April 2024
+.. |copyright-date| replace:: 2024
+.. |footer| replace:: Copyright © |copyright-date|, Arm Limited and its
+                      affiliates. All rights reserved.
+
+.. _ARMARM: https://developer.arm.com/documentation/ddi0487/latest
+.. _AAELF64: https://github.com/ARM-software/abi-aa/releases
+.. _CPPABI64: https://github.com/ARM-software/abi-aa/releases
+.. _CSTD: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf
+
+*********************************************************************************************
+C/C++ Atomics Application Binary Interface Standard for the Arm\ :sup:`®` 64-bit Architecture
+*********************************************************************************************
+
+.. class:: version
+
+|release|
+
+.. class:: issued
+
+Date of Issue: |date-of-issue|
+
+.. class:: logo
+
+.. image:: Arm_logo_blue_RGB.svg
+   :scale: 30%
+
+.. section-numbering::
+
+.. raw:: pdf
+
+   PageBreak oneColumn
+
+
+Preamble
+========
+
+Abstract
+--------
+
+This document describes the C/C++ Atomics Application Binary Interface for the
```

Comment:
I'm trying to work out what part of this document is ABI, i.e. what are the requirements for an implementation of the C/C++ atomic operations to be ABI compatible with other ABI compatible implementations; and what part is a cookbook of known/tested implementations. Which while helpful, I'm not sure is strictly part of the ABI?

I think it could be that any assembly sequence is considered ABI compatible if it passes Mix testing against the C memory model, and this document lists the sequences that have passed that bar already?

Other things I was expecting to see in an ABI document is whether a conforming implementation can use locks? For example I think one of the problems for ABI is whether an operation is lock-free or not. I think all the assembly sequences here are lock-free. If I'm right (not an expert in this area) then could be worth making it explicit.

Reply:
Chapter 4 is the ABI. Yes we have to list the various sequences plus explain the pitfalls because using atomics correctly is extremely hard. Compilers get these sequences wrong all the time without anyone noticing. So if you do not emit the listed sequences, then you're mostly likely getting atomics wrong.

Locks are inherently inefficient, unfair and can easily lead to deadlock. The 8/16/32/64/128-bit atomic sequences shown are obviously lock-free irrespectively of the architecture version. We can make that more explicit, including how to deal with the wacky non-power of 2 structure types that C/C++ allow.